contracts/Marketplace.sol

```diff
@@ -373,24 +376,43 @@ contract Marketplace is Proofs, StateRetrieval, Endian {
     address withdrawRecipient
   ) public {
     Request storage request = _requests[requestId];
-    require(
-      block.timestamp > requestExpiry(requestId),
-      "Request not yet timed out"
-    );
     require(request.client == msg.sender, "Invalid client address");
     RequestContext storage context = _requestContexts[requestId];
-    require(context.state == RequestState.New, "Invalid state");
+    RequestState state = requestState(requestId);
+    require(
+      state == RequestState.Cancelled ||
+        state == RequestState.Failed ||
+        state == RequestState.Finished,
+      "Invalid state"
+    );
+    require(context.fundsToReturnToClient != 0, "Nothing to withdraw");
+
+    if (state == RequestState.Cancelled) {
+      context.state = RequestState.Cancelled;
+      emit RequestCancelled(requestId);
+
+      // `fundsToReturnToClient` currently tracks funds to be returned for requests that successfully finish.
+      // When requests are cancelled, funds earmarked for payment for the duration
+      // between request expiry and request end (for every slot that was filled), should be returned to the client.
+      // Update `fundsToReturnToClient` to reflect this.
+      context.fundsToReturnToClient +=
+        context.slotsFilled *
+        _payoutAmount(requestId, requestExpiry(requestId));
+    } else if (state == RequestState.Failed) {
+      // For Failed requests the client is refunded whole amount.
+      context.fundsToReturnToClient = request.maxPrice();
+    } else {
+      context.state = RequestState.Finished;
+    }
 
-    // Update request state to Cancelled. Handle in the withdraw transaction
-    // as there needs to be someone to pay for the gas to update the state
-    context.state = RequestState.Cancelled;
     _removeFromMyRequests(request.client, requestId);
 
-    emit RequestCancelled(requestId);
-
-    uint256 amount = context.expiryFundsWithdraw;
+    uint256 amount = context.fundsToReturnToClient;
     _marketplaceTotals.sent += amount;
     assert(_token.transfer(withdrawRecipient, amount));
+
+    // We zero out the funds tracking in order to prevent double-spends
+    context.fundsToReturnToClient = 0;
```

Comment:
I don't think it is necessary. Also the reason why I would avoid this approach is because we will lose information about how the request terminated if we move from the current terminating states into "one that catches them all".

Reply:
Good  point, we still need the distinction on how a request ended to handle `freeSlot()` correctly.